### PR TITLE
Problem: pulpcore does not have pull-through cache feature

### PR DESCRIPTION
**Pull request by dkliban**
Solution: extend the content app to support pull-through cache

This patch enables the content app handler to look for a remote on a distribution. When a remote
is associated with a distribution, that remote is used to download content missing from Pulp.

re: #3894
https://pulp.plan.io/issues/3894

### CODE REVIEW

**Comment by dkliban** [comment]
@goosemania I had to close the original PR because Travis was very confused after my rebase. I re-opened and it's passing again. Please re-review.

File: pulpcore/content/handler.py [hunk 1]
```diff
@@ -301,8 +295,37 @@ def _save_content_artifact(self, download_result, content_artifact):
                     artifact.save()
             except IntegrityError:
                 artifact = Artifact.objects.get(artifact.q())
-            content_artifact.artifact = artifact
-            content_artifact.save()
+            update_content_artifact = True
+            if not content_artifact.pk:
```
Comment (daviddavis) [inline pulpcore/content/handler.py]: `content_artifact.pk` will always be set now that we are using UUIDs. You probably need to use `content_artifact._state.adding` here.

Comment (dkliban) [reply pulpcore/content/handler.py]: Thanks! I've updated and pushed again.

File: pulpcore/app/migrations/0002_distribution_remote.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('pulp_app', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='distribution',
+            name='remote',
+            field=models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='distributions', to='pulp_app.Remote'),
```
Comment (mdellweg) [inline pulpcore/app/migrations/0002_distribution_remote.py]: @dkliban This filed breaks the pulp_deb plugin in it's current implementation.
Should pulpcore fields not be named with `_` here?

Comment (dkliban) [reply pulpcore/app/migrations/0002_distribution_remote.py]: @mdellweg How does it break pulp_deb? I don't see a problem with the field being called 'remote' vs '_remote'.

Comment (mdellweg) [reply pulpcore/app/migrations/0002_distribution_remote.py]: Sorry, not explicit enough. The field is called `distributions` and that is a conflict.
`remote` is the class, i think.

Comment (mdellweg) [reply pulpcore/app/migrations/0002_distribution_remote.py]: Oh, i start to see, that the distributions is the autogenerated other side of the one2many relation. Anyway, it is a problem, that this name has been taken.

**Pull request merged**